nbs_probabilistic/create_dropout_ensemble.py

```diff
@@ -0,0 +1,145 @@
+    reduce_lr_patience=args['reduce_lr_patience']
+    data_subsample=args['data_subsample']
+    norm_subsample=args['norm_subsample']
+    bn_position=args['bn_position']
+    dropout=args['dropout']
+    l2=args['l2']
+    nt_in=args['nt_in'] #Ques: Is nt and nt_in the same thing? important since it casuses problem in input size for model layer 1 sometimes.
+    #nt_in=args['nt']
+    activation=args['activation']
+    min_es_delta=args['min_es_delta'] #not for all configs.
+    
+    
+    #run data generator
+    ds = xr.merge([xr.open_mfdataset(f'../../data/WeatherBench/5.625deg/{var}/*.nc',
+                                     combine='by_coords') for var in var_dict.keys()])
+    mean = xr.open_dataarray(f'../../data/WeatherBench/predictions/saved_models/{exp_id}_mean.nc') 
+    std = xr.open_dataarray(f'../../data/WeatherBench/predictions/saved_models/{exp_id}_std.nc')
+    
+    ds_test= ds.sel(time=slice(test_years[0],test_years[-1]))
+
+    #Question: Should we input data_subsample, norm_subsample, nt_in (or nt?), dt_in?
+    #Should shuffle be False? since its testing.
+    dg_test = DataGenerator(ds_test, var_dict, lead_time, batch_size=batch_size, shuffle=False,
+                            load=True,mean=mean, std=std, output_vars=output_vars,
+                            data_subsample=data_subsample, norm_subsample=norm_subsample,
+                            nt_in=nt_in,dt_in=dt_in) 
+    return dg_test
+
+def get_model(exp_id):
+    tf.compat.v1.disable_eager_execution() #needed?
+    saved_model_path=f'../../data/WeatherBench/predictions/saved_models/{exp_id}.h5'
+    
+    #ToDo: make loss function general.
+    #Break the code with error msg if not resnet or unet_google
+    #Important Ques: Since we dont build again, are we passing kernel, filters, activation, dropout and other details to the network?
+    if 'resnet' in exp_id:
+        model=tf.keras.models.load_model(saved_model_path
+                ,custom_objects={'PeriodicConv2D':PeriodicConv2D,'lat_mse': tf.keras.losses.mse})
+    elif 'unet_google' in exp_id:
+        model=tf.keras.models.load_model(saved_model_path, 
+                                         custom_objects={'PeriodicConv2D':PeriodicConv2D})#loss fn.?
+    else:
+        print('Only resnet or unet_google allowed')
+        #ToDO: Raise Exception
+        #Ques: Does Unet (not unet_google) work?
+    #model.summary()
+    return model
+
+def predict(dg_test,model,number_of_forecasts):
+    X,y=dg_test[0] #currently limiting output due to RAM issues.
+#     for i in range(len(dg_test)-1):
+#     X_batch,y_batch=dg_test[i+1]
+#     X=np.append(X,X_batch,axis=0)
+#     y=np.append(y,y_batch,axis=0)
+    
+    #test-time dropout
+    func = K.function(model.inputs + [K.learning_phase()], model.outputs)
+    pred_ensemble = np.array([np.asarray(func([X] + [1.]), dtype=np.float32).squeeze() for _ in
+                              range(number_of_forecasts)])
```

Comment:
I assume the GPU RAM is the problem?

The problem could be that you are loading the entire dataset and then ask the network to make a prediction all at once. It would be better to create predictions for each batch and then concatenate the predictions afterwards.

Here is some pseudocode
```
preds = []
for X, y in dg_test:
    preds.append(np.array([np.asarray(func([X] + [1.]), dtype=np.float32).squeeze() for _ in
                              range(number_of_forecasts)]))
pred_ensemble = np.array(preds)
```
I think something like this should work and should use a lot less GPU memory.

Reply:
The kernel does not get killed, but still takes too long!
**Can you tell me how to work with a GPU?**
I tried running
print("Num GPUs Available: ", len(tf.config.experimental.list_physical_devices('GPU')))
and it says 0.

Do I need to install tensorflow-gpu?

Reply:
Yes, you need to explicitly install tensorflow-gpu. So run cunda uninstall tensorflow, then conda install tensorflow-gpu. Tahsn should work.

Reply:
ok. I installed tensorflow-gpu=2.0.0 using Conda-forge. I didn't use 2.1 because the CUDA version (10.0) would not have been compatible.
everything seems to be fine. you can also check in the new commit to verify.

Now I am trying to run the code for 100 forecasts, it looks that **it still will take around 3 hours. Is that reasonable?**
How should I run the code overnight so that it doesn't terminate abruptly? 
I haven't used screen yet, but I'll do that if it helps. Anything else I should ensure?

Reply:
tf 2.0 is fine. 

3 hours still seems long. When you are done working on this, can you close your notebook and let me know? Then I can log into the PC and test it out. 

Yes, screen would fix that issue.

Reply:
I'm still on iteration 50 of 250+, so if you wish I can close it now? you can then check 01-test-time-dropout

Reply:
Ok, I will look at it tomorrow morning.

Reply:
Ok. then I'll let it run for now. thanks